File: app/frontend/src/api/methods.ts
/** methods requests and models; moved without changing payloads or responses. */
import { request, requestRaw } from './client'

export type MethodPrintSettings = {
  default_printer: string
  paper: 'A4' | 'A3' | 'Letter'
  orientation: 'portrait' | 'landscape'
  margin_top_mm: number
  margin_right_mm: number
  margin_bottom_mm: number
  margin_left_mm: number
  layout: 'standard' | 'compact'
  font_size_pt: number
  copies: number
  duplex: 'none' | 'long_edge' | 'short_edge'
  color: boolean
  preview_before_print: boolean
}

export type PrinterOption = {
  name: string
  display_name: string
  virtual: boolean
  system: boolean
  default: boolean
}

export type PrintJob = {
  id: string
  method_id: number
  method_version: number
  printer_name: string
  status: 'rendered' | 'queued' | 'completed' | 'failed'
  pdf_path: string
  output_path: string | null
  page_count: number
  field_count: number
  error_code: string | null
  error_message: string | null
  created_at: string
}

export type MethodValidationIssue = {
  field: string
  code: string
  message: string
}

export type AngleExposure = {
  angle_deg: number
  storage_mode: 'averaged' | 'full_interval'
  start_frame: number
  end_frame: number
}

export type MethodConditions = {
  ccd_layout_id: string | number
  selected_ccds: number[]
  dispersion_calibration_id: string | number
  reference_wavelength_nm: number
  actual_reference_wavelength_nm: number
  reference_width_points: number
  analysis_unit: 'ug/g' | 'mg/g' | '%'
  calculation_profile: 'legacy_2_0_2' | 'modern_v1'
  pre_excitation_seconds: number
  sampling_period_seconds: number
  frame_count: number
  dark_frame_count: number
  sample_repeats: number
  standard_repeats: number
  control_repeats: number
  standard_sample_name: string
  maximum_id_deviation: number
  rsd_enabled: boolean
  rsd_threshold: number
  calibration_threshold: number
  qc_threshold: number
  abnormal_threshold: number
  angle_exposures: AngleExposure[]
  storage_profile?: string
}

export type MethodVersion = {
  id: number
  version: number
  state: 'draft' | 'published'
  conditions: MethodConditions
  lines: SpectralLine[]
  validation_errors: MethodValidationIssue[]
  content_sha256: string
  created_at: string
}

export type StandardPoint = {
  name: string
  value: number
  active: boolean
}

export type LineDetectability = {
  detectable: boolean
  reason_code: string
  message: string
  ccd_index?: number
  ccd_label?: string
  point_index?: number
  frame_index?: number
  angle_slot?: number
  angle_deg?: number | null
}

export type SpectralLineInput = {
  line_type: 'baseline' | 'analysis' | 'internal_standard' | 'positioning'
  element: string
  wavelength_nm: number
  actual_wavelength_nm: number | null
  enabled: boolean
  critical_band: boolean
  priority: number
  background_line_id: string | null
  alignment_line_id: string | null
  internal_standard_mode: 'none' | 'background' | 'line'
  internal_standard_line_id: string | null
  scan_width_points: number
  background_offset_points: number
  peak_mode: 'max_single_point' | 'gaussian'
  peak_width_points: number
  fit_mode: 'linear' | 'quadratic' | 'cubic' | 'spline'
  coordinate_type: 'normal' | 'logarithmic'
  unit: 'ug/g' | 'mg/g' | '%'
  value_kind: 'content' | 'concentration'
  decimal_places: number
  lower_peak: number
  minimum_peak_ratio: number
  valid_range_min: number
  valid_range_max: number
  over_limit_tolerance_percent: number
  standard_points: StandardPoint[]
}

export type SpectralLine = SpectralLineInput & {
  id: string
  order: number
  reference_baseline: boolean
  detectability?: LineDetectability
}

export type SpectralLineOptions = {
  element_symbols: string[]
  line_types: Array<{ value: SpectralLineInput['line_type']; label: string }>
  internal_standard_modes: Array<{ value: SpectralLineInput['internal_standard_mode']; label: string }>
  peak_modes: Array<{ value: SpectralLineInput['peak_mode']; label: string }>
  fit_modes: Array<{ value: SpectralLineInput['fit_mode']; label: string }>
  coordinate_types: Array<{ value: SpectralLineInput['coordinate_type']; label: string }>
  limits: Record<string, number[] | number>
}

export type MethodLineCollection = {
  method_id: number
  version: number
  state: 'draft' | 'published'
  lines: SpectralLine[]
}

export type MethodRecord = {
  id: number
  name: string
  description: string
  work_type: string
  status: 'active' | 'paused' | 'deleted'
  current_version: number | null
  latest_version: number | null
  version: MethodVersion | null
  published_version: MethodVersion | null
  is_current: boolean
  created_at: string
  updated_at: string
}

export type CurrentMethodState = {
  method_id: number | null
  version: number | null
  work_type: string | null
  title: string | null
  status: string | null
  action_state: string
  actions: Record<string, boolean>
  method: MethodRecord | null
  referenced_version: MethodVersion | null
}

export type CcdLayoutOption = {
  id: number
  name: string
  frame_count: number
  ccds_per_frame: number
  points_per_ccd: number
  point_width_um: number
  allow_drift_um: number
  ccd_indices: number[]
  ccd_labels: string[]
}

export type DispersionOption = {
  id: number
  name: string
  ccd_layout_id: number
  wavelength_min: number
  wavelength_max: number
  enabled: boolean
  ccd_ranges: Array<{
    ccd_index: number
    wavelength_start_nm: number
    wavelength_end_nm: number
    safe_start_nm: number
    safe_end_nm: number
  }>
}

export type MethodOptions = {
  ccd_layouts: CcdLayoutOption[]
  dispersion_calibrations: DispersionOption[]
  storage_modes: Array<{ value: AngleExposure['storage_mode']; label: string }>
  limits: Record<string, number[] | number>
}

export const methodsApi = {
  methods: (token: string, includeDeleted = false) => request<MethodRecord[]>(`/api/v1/methods${includeDeleted ? '?include_deleted=true' : ''}`, { headers: { Authorization: `Bearer ${token}` } }),
  methodOptions: (token: string) => request<MethodOptions>('/api/v1/methods/options', { headers: { Authorization: `Bearer ${token}` } }),
  currentMethod: (token: string) => request<CurrentMethodState>('/api/v1/methods/current', { headers: { Authorization: `Bearer ${token}` } }),
  createMethod: (token: string, payload: { name: string; description?: string; work_type?: string }) => request<MethodRecord>('/api/v1/methods', { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  updateMethod: (token: string, methodId: number, payload: { name?: string; description?: string; work_type?: string; conditions?: MethodConditions }) => request<MethodRecord>(`/api/v1/methods/${methodId}`, { method: 'PATCH', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  publishMethod: (token: string, methodId: number) => request<MethodRecord>(`/api/v1/methods/${methodId}/publish`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  copyMethod: (token: string, methodId: number, name: string) => request<MethodRecord>(`/api/v1/methods/${methodId}/copy`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ name }) }),
  openMethod: (token: string, methodId: number) => request<MethodRecord>(`/api/v1/methods/${methodId}/open`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  pauseMethod: (token: string, methodId: number) => request<MethodRecord>(`/api/v1/methods/${methodId}/pause`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  resumeMethod: (token: string, methodId: number) => request<MethodRecord>(`/api/v1/methods/${methodId}/resume`, { method: 'POST', headers: { Authorization: `Bearer ${token}` } }),
  deleteMethod: (token: string, methodId: number) => request<{ id: number; deleted: boolean }>(`/api/v1/methods/${methodId}`, { method: 'DELETE', headers: { Authorization: `Bearer ${token}` } }),
  spectralLineOptions: (token: string) => request<SpectralLineOptions>('/api/v1/spectral-lines/options', { headers: { Authorization: `Bearer ${token}` } }),
  methodLines: (token: string, methodId: number) => request<MethodLineCollection>(`/api/v1/methods/${methodId}/lines`, { headers: { Authorization: `Bearer ${token}` } }),
  detectLine: (token: string, methodId: number, payload: { wavelength_nm: number; actual_wavelength_nm?: number | null; scan_width_points?: number }) => request<LineDetectability>(`/api/v1/methods/${methodId}/lines/detect`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  createLine: (token: string, methodId: number, payload: SpectralLineInput) => request<MethodRecord>(`/api/v1/methods/${methodId}/lines`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  updateLine: (token: string, methodId: number, lineId: string, payload: SpectralLineInput) => request<MethodRecord>(`/api/v1/methods/${methodId}/lines/${lineId}`, { method: 'PATCH', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(payload) }),
  deleteLine: (token: string, methodId: number, lineId: string) => request<MethodRecord>(`/api/v1/methods/${methodId}/lines/${lineId}`, { method: 'DELETE', headers: { Authorization: `Bearer ${token}` } }),
  reorderLines: (token: string, methodId: number, lineIds: string[]) => request<MethodRecord>(`/api/v1/methods/${methodId}/lines/reorder`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ line_ids: lineIds }) }),
  methodPrintSettings: (token: string) => request<MethodPrintSettings>('/api/v1/method-print/settings', { headers: { Authorization: `Bearer ${token}` } }),
  saveMethodPrintSettings: (token: string, settings: MethodPrintSettings) => request<MethodPrintSettings>('/api/v1/method-print/settings', { method: 'PATCH', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify(settings) }),
  methodPrinters: (token: string) => request<{ printers: PrinterOption[] }>('/api/v1/method-print/printers', { headers: { Authorization: `Bearer ${token}` } }),
  methodPreview: async (token: string, methodId: number, version: number | null, settings: MethodPrintSettings) => {
    const response = await requestRaw(`/api/v1/methods/${methodId}/preview`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ version, settings }) })
    return {
      html: await response.text(),
      pageCount: Number(response.headers.get('X-Page-Count') ?? 0),
      fieldCount: Number(response.headers.get('X-Field-Count') ?? 0),
      version: Number(response.headers.get('X-Method-Version') ?? version ?? 0),
    }
  },
  methodPdf: async (token: string, methodId: number, version: number | null, settings: MethodPrintSettings) => {
    const response = await requestRaw(`/api/v1/methods/${methodId}/pdf`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ version, settings }) })
    return {
      blob: await response.blob(),
      pageCount: Number(response.headers.get('X-Page-Count') ?? 0),
      fieldCount: Number(response.headers.get('X-Field-Count') ?? 0),
    }
  },
  printMethod: (token: string, methodId: number, version: number | null, settings: MethodPrintSettings, printerName: string) => request<PrintJob>(`/api/v1/methods/${methodId}/print`, { method: 'POST', headers: { Authorization: `Bearer ${token}` }, body: JSON.stringify({ version, settings, printer_name: printerName }) }),
  methodPrintJobs: (token: string, methodId: number) => request<{ jobs: PrintJob[] }>(`/api/v1/methods/${methodId}/print-jobs`, { headers: { Authorization: `Bearer ${token}` } }),
}

File: app/backend/schemas/methods.py
"""Methods request and response models, moved without rule changes."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class MethodCondition(BaseModel):
    """Editable method conditions.

    Numeric bounds are deliberately checked by the method domain service rather
    than Pydantic.  A draft must be able to retain invalid values and return
    field-level errors until the operator fixes it before publishing.
    """

    model_config = ConfigDict(extra="allow")

    ccd_layout_id: str | int = "default"
    selected_ccds: list[int] = Field(default_factory=lambda: [0, 1, 2, 4, 5])
    dispersion_calibration_id: str | int = "default"
    reference_wavelength_nm: float = 253.65
    actual_reference_wavelength_nm: float = 253.65
    reference_width_points: int = 21
    analysis_unit: str = "ug/g"
    calculation_profile: Literal["legacy_2_0_2", "modern_v1"] = "modern_v1"
    pre_excitation_seconds: float = 3.0
    sampling_period_seconds: float = 1.0
    frame_count: int = 20
    dark_frame_count: int = 8
    sample_repeats: int = 1
    standard_repeats: int = 3
    control_repeats: int = 1
    standard_sample_name: str = Field(default="", max_length=100)
    maximum_id_deviation: float = 5.0
    rsd_enabled: bool = True
    rsd_threshold: float = 5.0
    calibration_threshold: float = 5.0
    qc_threshold: float = 0.0
    abnormal_threshold: float = 0.0
    angle_exposures: list[dict[str, Any]] = Field(
        default_factory=lambda: [
            {
                "angle_deg": 0.0,
                "storage_mode": "averaged",
                "start_frame": 1,
                "end_frame": 20,
            }
        ]
    )


class MethodCreate(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    description: str = Field(default="", max_length=500)
    work_type: str = Field(default="spectral", min_length=1, max_length=50)
    conditions: MethodCondition | dict[str, Any] = Field(default_factory=MethodCondition)


class MethodUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=500)
    work_type: str | None = Field(default=None, min_length=1, max_length=50)
    conditions: MethodCondition | dict[str, Any] | None = None


class MethodActionRequest(BaseModel):
    method_id: int | None = None


class MethodVersion(BaseModel):
    id: int
    version: int
    state: Literal["draft", "published"]
    conditions: dict[str, Any]
    lines: list[dict[str, Any]] = Field(default_factory=list)
    validation_errors: list[dict[str, Any]]
    content_sha256: str
    created_at: datetime


class StandardPointInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(default="", max_length=50)
    value: float
    active: bool = True


class SpectralLineInput(BaseModel):
    """Complete editable spectral-line payload.

    Domain bounds stay in the service so invalid references and numeric rules
    return stable S04 error codes instead of generic request parsing errors.
    """

    model_config = ConfigDict(extra="forbid")

    line_type: Literal["baseline", "analysis", "internal_standard", "positioning"] = "analysis"
    element: str = Field(min_length=1, max_length=20)
    wavelength_nm: float
    actual_wavelength_nm: float | None = None
    enabled: bool = True
    critical_band: bool = False
    priority: int = 0
    background_line_id: str | None = None
    alignment_line_id: str | None = None
    internal_standard_mode: Literal["none", "background", "line"] = "none"
    internal_standard_line_id: str | None = None
    scan_width_points: int = 9
    background_offset_points: int = 0
    peak_mode: Literal["max_single_point", "gaussian"] = "max_single_point"
    peak_width_points: int = 1
    fit_mode: Literal["linear", "quadratic", "cubic", "spline"] = "linear"
    coordinate_type: Literal["normal", "logarithmic"] = "normal"
    unit: Literal["ug/g", "mg/g", "%"] = "ug/g"
    value_kind: Literal["content", "concentration"] = "content"
    decimal_places: int = 2
    lower_peak: int = 300
    minimum_peak_ratio: float = 1.5
    valid_range_min: float = 0.0
    valid_range_max: float = 9_999_999.0
    over_limit_tolerance_percent: float = 0.0
    standard_points: list[StandardPointInput] = Field(default_factory=list)


class SpectralLineReorder(BaseModel):
    line_ids: list[str]


class SpectralLineDetectRequest(BaseModel):
    wavelength_nm: float
    actual_wavelength_nm: float | None = None
    scan_width_points: int = 9


class MethodPrintSettings(BaseModel):
    model_config = ConfigDict(extra="forbid")

    default_printer: str = Field(default="geospectrum-pdf", min_length=1, max_length=255)
    paper: Literal["A4", "A3", "Letter"] = "A4"
    orientation: Literal["portrait", "landscape"] = "portrait"
    margin_top_mm: float = Field(default=12, ge=5, le=40)
    margin_right_mm: float = Field(default=12, ge=5, le=40)
    margin_bottom_mm: float = Field(default=12, ge=5, le=40)
    margin_left_mm: float = Field(default=12, ge=5, le=40)
    layout: Literal["standard", "compact"] = "standard"
    font_size_pt: int = Field(default=9, ge=8, le=12)
    copies: int = Field(default=1, ge=1, le=99)
    duplex: Literal["none", "long_edge", "short_edge"] = "none"
    color: bool = False
    preview_before_print: bool = True


class MethodRenderRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    version: int | None = Field(default=None, ge=1)
    settings: MethodPrintSettings | None = None


class MethodPrintRequest(MethodRenderRequest):
    printer_name: str | None = Field(default=None, min_length=1, max_length=255)


class MethodResponse(BaseModel):
    id: int
    name: str
    description: str
    work_type: str
    status: str
    current_version: int | None
    latest_version: int | None
    version: MethodVersion | None = None
    published_version: MethodVersion | None = None
    is_current: bool = False
    created_at: datetime
    updated_at: datetime


class MethodCurrentResponse(BaseModel):
    method_id: int | None
    version: int | None
    work_type: str | None
    title: str | None
    status: str | None
    action_state: str
    actions: dict[str, bool]
    method: MethodResponse | None = None
    referenced_version: MethodVersion | None = None
